src/chunk.c

```diff
@@ -774,23 +775,24 @@ copy_hypertable_acl_to_relid(Hypertable *ht, Oid relid)
  * instead needs the proper permissions on the database to create the schema.
  */
 Oid
-ts_chunk_create_table(Chunk *chunk, Hypertable *ht, const char *tablespacename)
+ts_chunk_create_table(const Chunk *chunk, const Hypertable *ht, const char *tablespacename)
 {
 	Relation rel;
 	ObjectAddress objaddr;
 	int sec_ctx;
-	char *namespace = NameStr(ht->fd.schema_name);
-	char *hyper_name = NameStr(ht->fd.table_name);
+	const char *namespace = NameStr(ht->fd.schema_name);
+	const char *hyper_name = NameStr(ht->fd.table_name);
 
 	/*
 	 * The CreateForeignTableStmt embeds a regular CreateStmt, so we can use
 	 * it to create both regular and foreign tables
 	 */
 	CreateForeignTableStmt stmt = {
 		.base.type = T_CreateStmt,
-		.base.relation =
-			makeRangeVar(NameStr(chunk->fd.schema_name), NameStr(chunk->fd.table_name), 0),
-		.base.inhRelations = list_make1(makeRangeVar(namespace, hyper_name, 0)),
+		.base.relation = makeRangeVar(pstrdup(NameStr(chunk->fd.schema_name)),
+									  pstrdup(NameStr(chunk->fd.table_name)),
```

Comment:
The chunk is now `const` so we cannot reference a `const` string in `makeRangeVar` since PostgreSQL expects mutable strings.

In general, PostgreSQL often mutates planner nodes, so to be safe we shouldn't pass in pointers to strings that also exist in other objects (like the chunk in this case). The fact that the chunk is `const` here actually makes it obvious that we shouldn't pass it's name and schema strings to the planner node without copying them first. This is a good thing, IMO, and might prevent a bunch of subtle bugs.

Reply:
But adding `const` does not in itself make it necessary to copy data, and the implementation of `makeRangeVar` does not mutate the string. It is only non-`const` because the PG interface is not `const`-correct.

The only reasons I can see for copying the string is that:
- The source has a shorter life-time than the destination. Changing from non-`const` to `const` does not protect against this anyway, and since the code already works, this could not be a reason to copy the string.
- The destination need to be modified. This is what adding `const` protects against, but `makeRangeVar` does not modify the string, so there is no reason to copy the string. However, it does not copy the string either, it just assigns it to the `schema_name` field, but looking over the code I do not see any place where the schema name is actually modified or rewritten. Yes, it is quite often assigned a different value, but that does not mean that the original string is modified, which would be a reason for copying the string.

The danger with copying the string is that we suddenly start to allocate more memory for the strings, which we should not start to do as part of a patch that should just refactor the code.

Reply:
I don't think just looking at the "constructor" of an object, like `makeRangeVar` is enough to conclude that the object or its contained sub-objects won't be modified. You also need to consider the code paths where the object is actually used.

Yes, `makeRangeVar` doesn't mutate the string, but it is commonly used as an object/node in a planner tree, where there is no guarantee it won't be mutated. We often mutate planner trees and objects in our own code, which has also been a source of bugs in the past.

I see no danger with allocating memory for strings. Quite the opposite, I see more dangers with passing around references that you never know where or when they are modified. Heap allocation in PostgreSQL is cheap given the memory context functionality and the memory is freed automatically, so I see little reason to judiciously try to save on memory. The `makeRangeVar` itself is created in the current memory context, so it make sense to copy also the passed in strings so that the range var is a deep copy instead of a shallow one. I guess it takes non-const string pointers so that one can decide on whether a deep or shallow copy makes sense. Clearly, a fully independent object is much safer than passing around references.

Reply:
> I am actually not sure what you are proposing. Do you prefer that we cast away the `const` in order to make references in `makeRangeVar` just to save a tiny amount of heap memory?

No, to cast away the `const` and not try to fix a non-existing problem in a refactoring commit.

The lifetime of `stmt` extends to the end of the function. This means that the lifetime of fields of `stmt` also are to the end of the function. Since `chunk` is a parameter, the lifetime of `chunk` extends beyond the function, which means that anything accessed in the lifetime of `stmt` can also be accessed in the lifetime of `chunk`. This means that there is no reason to copy data from `chunk` to `stmt` since data available in `chunk` will exist for the duration of `stmt` (unless memory is deliberately deallocated in the function, which it is not).

Reply:
@mkindahl These changes does not address any lifecycle/lifetime issues, so not sure why that's a concern. These changes address issues related to mutability, not lifetime. Marking something as `const` means that that object should not be modified. Thus, passing that const object to a function that doesn't take a const version of the object as input implicitly means that the function can modify the passed in object. Therefore, as a rule, it makes sense to pass in a mutable copy and  not a reference to the object, since that is the only way to guarantee that the object is not mutated in the function, nor in a future update to that function.

Reply:
FWIW, removing the pstrdup leads to this compiler error:

```
/Users/enordstr/Projects/timescaledb/src/chunk.c:796:33: error: passing 'char const[64]' to parameter of type 'char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
                .base.relation = makeRangeVar(NameStr(chunk->fd.schema_name),

```

So, the only other options from making a pstrdup copy is to cast away the `const` or revert the change making `Chunk *` a `const Chunk *`. But the latter is the whole point of this PR: to constify objects passed around in the code.